polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -155,9 +158,78 @@ static bool isClassType(Type type, StringRef className) {
 }
 
 namespace {
-struct RaiseKernelName : public OpRewritePattern<LLVM::AddressOfOp> {
+template <typename SourceOp>
+class OpOrInterfaceRaisePatternBase : public RewritePattern {
+public:
+  using RewritePattern::RewritePattern;
+
+  /// Wrappers around the RewritePattern methods that pass the derived op type.
+  void rewrite(Operation *op, PatternRewriter &rewriter) const final {
+    rewrite(cast<SourceOp>(op), rewriter);
+  }
+  LogicalResult match(Operation *op) const final {
+    // Do not run raising patterns in device code
+    return isInDeviceModule(op) ? failure() : match(cast<SourceOp>(op));
+  }
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const final {
+    // Do not run raising patterns in device code
+    return isInDeviceModule(op) ? failure()
+                                : matchAndRewrite(cast<SourceOp>(op), rewriter);
+  }
+
+  /// Rewrite and Match methods that operate on the SourceOp type. These must be
+  /// overridden by the derived pattern class.
+  virtual void rewrite(SourceOp op, PatternRewriter &rewriter) const {
+    llvm_unreachable("must override rewrite or matchAndRewrite");
+  }
+  virtual LogicalResult match(SourceOp op) const {
+    llvm_unreachable("must override match or matchAndRewrite");
+  }
+  virtual LogicalResult matchAndRewrite(SourceOp op,
+                                        PatternRewriter &rewriter) const {
+    if (succeeded(match(op))) {
+      rewrite(op, rewriter);
+      return success();
+    }
+    return failure();
+  }
+
+private:
+  static bool isInDeviceModule(Operation *op) {
+    return op->getParentOfType<gpu::GPUModuleOp>();
+  }
+};
+
+/// OpRaisePattern is a wrapper around RaisePattern that allows for
+/// matching and rewriting against an instance of a derived operation class as
+/// opposed to a raw Operation.
+template <typename SourceOp>
+struct OpRaisePattern : public OpOrInterfaceRaisePatternBase<SourceOp> {
```

Comment:
Should we add `Host` prefix to the name here and below to make it clearer?

Reply:
Done